### PR TITLE
executor: change outer result to list in index merge join

### PR DESCRIPTION
**Pull request by lzmhhh123**
### What problem does this PR solve? 
As the title says. The outer result is changed to list to avoid fetch little data in the outer side and the memory copy.

Reference issue: https://github.com/pingcap/tidb/issues/12498

### What is changed and how it works?
First, add a vectorized filter function for list. Then change some iteration logic.

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

 - Has exported function/method change

Side effects

 - Increased code complexity

### CODE REVIEW

**Comment by SunRunAway** [comment]
Could you provide a benchmark result?

**Comment by lzmhhh123** [comment]
New benchmark compares with index hash join.
```bash
$ CGO_ENABLED=1 GO111MODULE=on go test -count 5 -benchmem -run=^$ github.com/pingcap/tidb/executor -bench BenchmarkIndexJoinExec
goos: darwin
goarch: amd64
pkg: github.com/pingcap/tidb/executor
BenchmarkIndexJoinExec/index_merge_join_need_outer_sort_(outerRows:100000,_innerRows:102400,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:true)-12                      8         223784819 ns/op        608316862 B/op    304097 allocs/op
BenchmarkIndexJoinExec/index_merge_join_need_outer_sort_(outerRows:100000,_innerRows:102400,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:true)-12                      8         258324156 ns/op        579590974 B/op    304086 allocs/op
BenchmarkIndexJoinExec/index_merge_join_need_outer_sort_(outerRows:100000,_innerRows:102400,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:true)-12                      7         294850408 ns/op        568235004 B/op    304058 allocs/op
BenchmarkIndexJoinExec/index_merge_join_need_outer_sort_(outerRows:100000,_innerRows:102400,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:true)-12                     12         296859596 ns/op        562721833 B/op    304040 allocs/op
BenchmarkIndexJoinExec/index_merge_join_need_outer_sort_(outerRows:100000,_innerRows:102400,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:true)-12                      9         287201640 ns/op        569117074 B/op    304046 allocs/op
BenchmarkIndexJoinExec/index_merge_join_(outerRows:100000,_innerRows:102400,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-12                                      7        262968188 ns/op        568202419 B/op    304019 allocs/op
BenchmarkIndexJoinExec/index_merge_join_(outerRows:100000,_innerRows:102400,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-12                                     7         293523160 ns/op        584710332 B/op    304062 allocs/op
BenchmarkIndexJoinExec/index_merge_join_(outerRows:100000,_innerRows:102400,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-12                                     6         262443727 ns/op        591506274 B/op    304046 allocs/op
BenchmarkIndexJoinExec/index_merge_join_(outerRows:100000,_innerRows:102400,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-12                                    13         330152695 ns/op        570100248 B/op    304022 allocs/op
BenchmarkIndexJoinExec/index_merge_join_(outerRows:100000,_innerRows:102400,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-12                                     3         346654902 ns/op        639537576 B/op    304115 allocs/op
BenchmarkIndexJoinExec/index_inner_hash_join_(outerRows:100000,_innerRows:102400,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-12                                1        1431682940 ns/op        2482053168 B/op   307324 allocs/op
BenchmarkIndexJoinExec/index_inner_hash_join_(outerRows:100000,_innerRows:102400,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-12                                1        1138102730 ns/op        2482030664 B/op   307157 allocs/op
BenchmarkIndexJoinExec/index_inner_hash_join_(outerRows:100000,_innerRows:102400,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-12                                2        1388877594 ns/op        2482049436 B/op   307301 allocs/op
BenchmarkIndexJoinExec/index_inner_hash_join_(outerRows:100000,_innerRows:102400,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-12                                3        1574714213 ns/op        2482024701 B/op   307216 allocs/op
BenchmarkIndexJoinExec/index_inner_hash_join_(outerRows:100000,_innerRows:102400,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-12                                1        1211453381 ns/op        2482039104 B/op   307232 allocs/op
BenchmarkIndexJoinExec/index_outer_hash_join_(outerRows:100000,_innerRows:102400,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-12                                2        1071477132 ns/op        2472659020 B/op   208629 allocs/op
BenchmarkIndexJoinExec/index_outer_hash_join_(outerRows:100000,_innerRows:102400,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-12                                2        1199482944 ns/op        2472663360 B/op   208657 allocs/op
BenchmarkIndexJoinExec/index_outer_hash_join_(outerRows:100000,_innerRows:102400,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-12                                2        1249291263 ns/op        2472656300 B/op   208608 allocs/op
BenchmarkIndexJoinExec/index_outer_hash_join_(outerRows:100000,_innerRows:102400,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-12                                3        1540364218 ns/op        2472658824 B/op   208621 allocs/op
BenchmarkIndexJoinExec/index_outer_hash_join_(outerRows:100000,_innerRows:102400,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-12                                1        1370140536 ns/op        2472659128 B/op   208627 allocs/op
PASS
ok      github.com/pingcap/tidb/executor        115.958s
```

File: executor/index_lookup_merge_join.go [hunk 1]
```diff
@@ -332,29 +332,35 @@ func (omw *outerMergeWorker) pushToChan(ctx context.Context, task *lookUpMergeJo
...
+		}
 
-	task.memTracker.Consume(task.outerResult.MemoryUsage())
-	oldMemUsage := task.outerResult.MemoryUsage()
-	err := Next(ctx, omw.executor, task.outerResult)
-	if err != nil {
-		return task, err
+		task.outerResult.Add(execChk)
+		requiredRows -= execChk.NumRows()
+		task.memTracker.Consume(execChk.MemoryUsage())
```
[the remark this answers is not in the document]
Comment (lzmhhh123) [reply executor/index_lookup_merge_join.go]: No, we can't. The `memTracker` is attached to the executor's `memTracker`.

[the remark this answers is not in the document]
Comment (SunRunAway) [reply executor/index_lookup_merge_join.go]: We can do `oldTask.memTracker.Consume(-task.memTracker.BytesConsumed())` in `getFinishedTask` when the old task is abadoned.

[the remark this answers is not in the document]
Comment (lzmhhh123) [reply executor/index_lookup_merge_join.go]: That should be changed with `index hash join` at the same time. I'd prefer to do it in another PR.

[the remark this answers is not in the document]
Comment (SunRunAway) [reply executor/index_lookup_merge_join.go]: Ok, Please create an issue.

**Comment by lzmhhh123** [comment]
Compare with the master version, hacked to chunk append like index join.

```bash
# liuzhuomin @ Charmings-Macbook-Pro in ~/.gvm/pkgsets/go1.13/global/src/github.com/pingcap/tidb on git:dev/index_merge_join_outerlist x [14:17:15]
$ benchstat old.txt new.txt
name                                                                                                                                                                      old time/op    new time/op    delta
IndexJoinExec/index_merge_join_need_outer_sort_(outerRows:100000,_innerRows:102400,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:true)-12     1.27s ±46%     0.26s ±20%  -79.32%  (p=0.008 n=5+5)
IndexJoinExec/index_merge_join_(outerRows:100000,_innerRows:102400,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-12                    1.26s ±45%     0.27s ±17%  -78.87%  (p=0.008 n=5+5)
IndexJoinExec/index_inner_hash_join_(outerRows:100000,_innerRows:102400,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-12               1.14s ±35%     1.17s ±32%     ~     (p=1.000 n=5+5)
IndexJoinExec/index_outer_hash_join_(outerRows:100000,_innerRows:102400,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-12               1.39s ±33%     1.19s ±26%     ~     (p=0.421 n=5+5)

name                                                                                                                                                                      old alloc/op   new alloc/op   delta
IndexJoinExec/index_merge_join_need_outer_sort_(outerRows:100000,_innerRows:102400,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:true)-12    3.38GB ± 0%    0.55GB ± 6%  -83.77%  (p=0.008 n=5+5)
IndexJoinExec/index_merge_join_(outerRows:100000,_innerRows:102400,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-12                   3.38GB ± 0%    0.59GB ± 6%  -82.67%  (p=0.008 n=5+5)
IndexJoinExec/index_inner_hash_join_(outerRows:100000,_innerRows:102400,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-12              2.48GB ± 0%    2.48GB ± 0%     ~     (p=1.000 n=5+5)
IndexJoinExec/index_outer_hash_join_(outerRows:100000,_innerRows:102400,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-12              2.47GB ± 0%    2.47GB ± 0%     ~     (p=0.056 n=5+5)

name                                                                                                                                                                      old allocs/op  new allocs/op  delta
IndexJoinExec/index_merge_join_need_outer_sort_(outerRows:100000,_innerRows:102400,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:true)-12      304k ± 0%      304k ± 0%     ~     (p=0.063 n=5+5)
IndexJoinExec/index_merge_join_(outerRows:100000,_innerRows:102400,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-12                     304k ± 0%      304k ± 0%   +0.06%  (p=0.008 n=5+5)
IndexJoinExec/index_inner_hash_join_(outerRows:100000,_innerRows:102400,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-12                307k ± 0%      307k ± 0%     ~     (p=1.000 n=5+5)
IndexJoinExec/index_outer_hash_join_(outerRows:100000,_innerRows:102400,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-12                209k ± 0%      209k ± 0%     ~     (p=0.087 n=5+5)
```

File: executor/index_lookup_merge_join.go [hunk 2]
```diff
@@ -394,20 +400,27 @@ func (imw *innerMergeWorker) run(ctx context.Context, wg *sync.WaitGroup, cancel
...
+			if err != nil {
+				return err
+			}
 		}
 	}
-	task.outerOrderIdx = make([]int, 0, numOuterRows)
-	for i := 0; i < numOuterRows; i++ {
-		if len(outerMatch) == 0 || outerMatch[i] {
-			task.outerOrderIdx = append(task.outerOrderIdx, i)
+	task.outerOrderIdx = make([]chunk.RowPtr, 0, numOuterRows)
```
Comment (XuHuaiyu) [inline executor/index_lookup_merge_join.go]: `numOuterRows` is only used here, so we can use `task.outerOrderIdx = make([]chunk.RowPtr, 0, task.outerResult.Len())` here

File: executor/index_lookup_merge_join.go [hunk 3]
```diff
@@ -413,13 +413,12 @@ func (imw *innerMergeWorker) handleTask(ctx context.Context, task *lookUpMergeJo
...
 		}
 	}
-	task.outerOrderIdx = make([]chunk.RowPtr, 0, numOuterRows)
-	numChunk := uint32(task.outerResult.NumChunks())
-	for i := uint32(0); i < numChunk; i++ {
-		numRow := uint32(task.outerResult.GetChunk(int(i)).NumRows())
-		for j := uint32(0); j < numRow; j++ {
+	task.outerOrderIdx = make([]chunk.RowPtr, 0, task.outerResult.Len())
+	for i := 0; i < numOuterChks; i++ {
+		numRow := task.outerResult.GetChunk(int(i)).NumRows()
```
Comment (XuHuaiyu) [inline executor/index_lookup_merge_join.go]: no need to use `int(i)`

**Pull request merged**